### PR TITLE
Shortcut simple patterns ending in `*`

### PR DESCRIPTION
**Pull request by DaveCTurner**
When profiling a call to `AllocationService#reroute()` in a large cluster
containing allocation filters of the form `node-name-*` I observed a nontrivial
amount of time spent in `Regex#simpleMatch` due to these allocation filters.
Patterns ending in a wildcard are not uncommon, and this change treats them as
a special case in `Regex#simpleMatch` in order to shave a bit of time off this
calculation. It also uses `String#regionMatches()` to avoid an allocation in
the case that the pattern's only wildcard is at the start.

Microbenchmark results before this change:

    Result "org.elasticsearch.common.regex.RegexStartsWithBenchmark.performSimpleMatch":
      1113.839 ±(99.9%) 6.338 ns/op [Average]
      (min, avg, max) = (1102.388, 1113.839, 1135.783), stdev = 9.486
      CI (99.9%): [1107.502, 1120.177] (assumes normal distribution)

Microbenchmark results with this change applied:

    Result "org.elasticsearch.common.regex.RegexStartsWithBenchmark.performSimpleMatch":
      433.190 ±(99.9%) 0.644 ns/op [Average]
      (min, avg, max) = (431.518, 433.190, 435.456), stdev = 0.964
      CI (99.9%): [432.546, 433.833] (assumes normal distribution)

The microbenchmark in question was:

    @Fork(3)
    @Warmup(iterations = 10)
    @Measurement(iterations = 10)
    @BenchmarkMode(Mode.AverageTime)
    @OutputTimeUnit(TimeUnit.NANOSECONDS)
    @State(Scope.Benchmark)
    @SuppressWarnings("unused") //invoked by benchmarking framework
    public class RegexStartsWithBenchmark {

        private static final String testString = "abcdefghijklmnopqrstuvwxyz";
        private static final String[] patterns;

        static {
            patterns = new String[testString.length() + 1];
            for (int i = 0; i <= testString.length(); i++) {
                patterns[i] = testString.substring(0, i) + "*";
            }
        }

        @Benchmark
        public void performSimpleMatch() {
            for (int i = 0; i < patterns.length; i++) {
                Regex.simpleMatch(patterns[i], testString);
            }
        }
    }

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/common/regex/Regex.java [hunk 1]
```diff
@@ -113,8 +113,12 @@ public static boolean simpleMatch(String pattern, String str) {
             }
             return false;
         }
+        final String patternBeforeFirstIndex = pattern.substring(0, firstIndex);
+        if (firstIndex == pattern.length() - 1) {
+            return str.startsWith(patternBeforeFirstIndex);
+        }
         return (str.length() >= firstIndex &&
-                pattern.substring(0, firstIndex).equals(str.substring(0, firstIndex)) &&
+                patternBeforeFirstIndex.equals(str.substring(0, firstIndex)) &&
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/common/regex/Regex.java]: While we are micro optimizing :), why not leave this as is and also inline the `pattern.substring(0, firstIndex)` inside the new conditional above? That way we still get the free breakout for short strings in the first part of this boolean?

File: server/src/test/java/org/elasticsearch/common/regex/RegexTests.java [hunk 2]
```diff
@@ -63,4 +63,24 @@ public void testDoubleWildcardMatch() {
         assertTrue(Regex.simpleMatch("fff*******ddd", "fffabcddd"));
         assertFalse(Regex.simpleMatch("fff******ddd", "fffabcdd"));
     }
+
+    public void testSimpleMatch() {
+        for (int i = 0; i < 100000; i++) {
```
Comment (original-brownbear) [inline server/src/test/java/org/elasticsearch/common/regex/RegexTests.java]: Leftover from benchmarking experiments?

Comment (DaveCTurner) [reply server/src/test/java/org/elasticsearch/common/regex/RegexTests.java]: Meh 100k iterations is still only 300ms :)

Comment (DaveCTurner) [reply server/src/test/java/org/elasticsearch/common/regex/RegexTests.java]: Reduced in 489e82b.

**Review by henningandersen — COMMENTED** [review]
Looking good, but I wonder if we could get rid of more allocations (if that was the culprit?), see comments.

File: server/src/main/java/org/elasticsearch/common/regex/Regex.java [hunk 3]
```diff
@@ -113,6 +113,9 @@ public static boolean simpleMatch(String pattern, String str) {
             }
             return false;
         }
+        if (firstIndex == pattern.length() - 1) {
+            return str.startsWith(pattern.substring(0, firstIndex));
```
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/common/regex/Regex.java]: TIL `regionMatches` is the name of the function I was trying to find. Yes, this shaves off a bit more time. I've updated the PR description with new benchmark results.

**Review by henningandersen — APPROVED** [review]
LGTM

**Review by original-brownbear — APPROVED** [review]
LGTM2

**Pull request merged**